codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -277,24 +282,27 @@ private int allocateBytesForTree(Http2Stream parent, int connectionWindow) {
 
         // This is the priority algorithm which will divide the available bytes based
         // upon stream weight relative to its peers
-        Http2Stream[] children = parent.children().toArray(new Http2Stream[parent.numChildren()]);
+        ChildCache childCache = new ChildCache(parent);
         int totalWeight = parent.totalChildWeights();
-        for (int tail = children.length; tail > 0;) {
-            int head = 0;
-            int nextTail = 0;
+        while (childCache.nextIterationSize() > 0) {
             int nextTotalWeight = 0;
             int nextConnectionWindow = connectionWindow;
-            for (; head < tail && nextConnectionWindow > 0; ++head) {
-                Http2Stream child = children[head];
+
+            for (Http2Stream child : childCache) {
```

Comment:
Maybe we could add a `Http2Stream forEachChild(StreamVisitor childVisitor) throws Http2Exception` to `Http2Stream` and remove the direct access to the `Collection<? extends Http2Stream> children();`.  The `StreamVisitor` we create would do the initial iteration via `forEachChild(visitor)` but it would also maintain state...which is the array for the next iteration.  This way we would only have 1 object allocation for each iteration...plus the interface exposure would be reduced.  WDTY?